datafusion/datasource/src/source.rs

```diff
@@ -261,11 +261,7 @@ impl ExecutionPlan for DataSourceExec {
     ) -> Result<SendableRecordBatchStream> {
         self.data_source
             .open(partition, Arc::clone(&context))
-            .map(|stream| wrap_yield_stream(stream, &context, self.cooperative))
-    }
-
-    fn with_cooperative_yields(self: Arc<Self>) -> Option<Arc<dyn ExecutionPlan>> {
-        self.cooperative.then_some(self)
+            .map(|stream| make_cooperative(stream))
```

Comment:
We could avoid a virtual call here by pushing the cooperative wrapper down to `FileScanConfig::open`. That does lose the benefit of `DataSourceExec` being cooperative for any `DataSource`. Is it worth shaving off one virtual call per record batch?

Reply:
That would make sense indeed. ExecutionPlan fulfills the role of a factory of Streams. DataSourceExec delegates that to the DataSource itself. So Datasource is actually the factory. It would make sense to delegate the 'will you create something cooperative?' query as well.